<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Thomas Quinn &amp; Ionas Erb" />

<meta name="date" content="2018-03-22" />

<title>Moderating Differential Proportionality</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Moderating Differential Proportionality</h1>
<h4 class="author"><em>Thomas Quinn &amp; Ionas Erb</em></h4>
<h4 class="date"><em>2018-03-22</em></h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This vignette reviews the <span class="math inline">\(\theta\)</span> types and their variant definitions. It also discusses how to calculate an <span class="math inline">\(F\)</span>-statistic from <span class="math inline">\(\theta_d\)</span>. In presenting this, we provide details on how to moderate the <span class="math inline">\(F\)</span>-statistic using <code>voom</code> from the <code>limma</code> package. To keep this vignette tractable yet reproducible, we use a subset of the <code>iris</code> data set as an example.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(propr)
<span class="kw">data</span>(iris)
keep &lt;-<span class="st"> </span>iris$Species %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;setosa&quot;</span>, <span class="st">&quot;versicolor&quot;</span>)
counts &lt;-<span class="st"> </span>iris[keep, <span class="dv">1</span>:<span class="dv">4</span>] *<span class="st"> </span><span class="dv">10</span>
group &lt;-<span class="st"> </span><span class="kw">ifelse</span>(iris[keep, <span class="st">&quot;Species&quot;</span>] ==<span class="st"> &quot;setosa&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>)</code></pre></div>
</div>
<div id="the-three-types-of-theta" class="section level2">
<h2>The three types of theta</h2>
<p>In the companion vignette, “Calculating Differential Proportionality”, we introduce three types of differential proportionality: <span class="math inline">\(\theta_d\)</span>, <span class="math inline">\(\theta_e\)</span>, and <span class="math inline">\(\theta_f\)</span>. Each of these are defined as a fraction of the within-group log-ratio variances (i.e., VLR) for two features compared with the total VLR. Although at least three types of <span class="math inline">\(\theta\)</span> exist, we focus here on <span class="math inline">\(\theta_d\)</span>. Given two groups, sized <span class="math inline">\(k\)</span> and <span class="math inline">\(n - k\)</span>, we define <span class="math inline">\(\theta_d\)</span>, a measure of <em>disjointed proportionality</em>, as the pooled (weighted) VLR within the two groups divided by the total VLR:</p>
<p><span class="math display">\[\theta_d(\textrm{X}, \textrm{Y}) = \frac{(k-1)\textrm{VLR}_1 + (n-k-1)\textrm{VLR}_2}{(n-1)\textrm{VLR}}\]</span></p>
</div>
<div id="the-four-variant-states-of-theta" class="section level2">
<h2>The four variant states of theta</h2>
<p>Recall from the companion vignette that we can approximate VLR using a power transformation based on a parameter <span class="math inline">\(\alpha\)</span> that is usually close to zero. There, we suggested that we could calculate any <span class="math inline">\(\theta\)</span> using the <span class="math inline">\(\alpha\)</span>-transformed VLR (i.e., <span class="math inline">\(\textrm{aVLR}\)</span>) in place of the VLR. Hence, we realize the first variant state of <span class="math inline">\(\theta_d\)</span>:</p>
<p><span class="math display">\[\theta_d(\textrm{X}, \textrm{Y}) = \frac{(k-1)\textrm{aVLR}_1 + (n-k-1)\textrm{aVLR}_2}{(n-1)\textrm{aVLR}}\]</span></p>
<p>Meanwhile, the third and fourth variant states of <span class="math inline">\(\theta\)</span> make use of precision weights as estimated by the <code>voom</code> function of the <code>limma</code> package. The nature of sequencing count data makes it such highly abundant genes have more variance than lowly abundant genes. However, this trend gets reversed with log-transformation: highly abundant genes come to have less variance after transformation. Therefore, it follows that we would systematically see less variance around log-ratio pairs involving highly abundant genes. We can thus use precision weights to correct this, placing all VLR measurements on the same scale.</p>
<p>The particulars of <code>voom</code> are established elsewhere, but we mention here that it returns a matrix, <span class="math inline">\(V\)</span>, of the per-feature weights for each sample. From this, we can calculate the weights of a feature ratio, <span class="math inline">\(\mathbf{V^{\textrm{X},\textrm{Y}}}\)</span>, as the element-wise product of the per-feature weights (i.e., <span class="math inline">\(\mathbf{V^{\textrm{X},\textrm{Y}}} = \mathbf{V}^\textrm{X} * \mathbf{V}^\textrm{Y}\)</span>). Analogous to other formulations of <span class="math inline">\(\theta\)</span>, this uses the weighted variance of the log-ratios (i.e., <span class="math inline">\(\textrm{wVLR}\)</span>) in place of the VLR. In addition, we use here the modifier <span class="math inline">\(\Omega=\sum \mathbf{w} - \frac{\sum \mathbf{w}^2}{\sum \mathbf{w}}\)</span> in place of the sample sizes, where <span class="math inline">\(\mathbf{w} = \mathbf{V^{\textrm{X},\textrm{Y}}}\)</span>:</p>
<p><span class="math display">\[\theta_d(\textrm{X}, \textrm{Y}) = \frac{(\Omega_1)\textrm{wVLR}_1 + (\Omega_2)\textrm{wVLR}_2}{(\Omega)\textrm{wVLR}}\]</span></p>
<p>Likewise, we can calculate <span class="math inline">\(\theta\)</span> using both precision weights and the <span class="math inline">\(\alpha\)</span> transformation together. This requires a weighted and <span class="math inline">\(\alpha\)</span>-transformed approximation of the VLR (i.e., <span class="math inline">\(\textrm{waVLR}\)</span>) to use in conjunction with the <span class="math inline">\(\Omega\)</span> modifier from above:</p>
<p><span class="math display">\[\theta_d(\textrm{X}, \textrm{Y}) = \frac{(\Omega_1)\textrm{waVLR}_1 + (\Omega_2)\textrm{waVLR}_2}{(\Omega)\textrm{waVLR}}\]</span></p>
<p>When using the <code>propd</code> function, it is simple to calculate any variant state of <span class="math inline">\(\theta\)</span> To use the <span class="math inline">\(\alpha\)</span>-transformation, set the <code>alpha</code> argument to any positive number. To use precision weights, set the <code>weighted</code> argument to <code>TRUE</code>. Note that changing either of these arguments affects all <span class="math inline">\(\theta\)</span> calculations, not just <span class="math inline">\(\theta_d\)</span>. The user can access the intermediate precision weights via the <code>@weights</code> slot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pd.nn &lt;-<span class="st"> </span><span class="kw">propd</span>(counts, group, <span class="dt">weighted =</span> <span class="ot">FALSE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pd.wn &lt;-<span class="st"> </span><span class="kw">propd</span>(counts, group, <span class="dt">weighted =</span> <span class="ot">TRUE</span>)
pd.na &lt;-<span class="st"> </span><span class="kw">propd</span>(counts, group, <span class="dt">weighted =</span> <span class="ot">FALSE</span>, <span class="dt">alpha =</span> .<span class="dv">01</span>)
pd.wa &lt;-<span class="st"> </span><span class="kw">propd</span>(counts, group, <span class="dt">weighted =</span> <span class="ot">TRUE</span>, <span class="dt">alpha =</span> .<span class="dv">01</span>)</code></pre></div>
<p>We refer the reader to Erb et al. 2017 for an elaboration of the four variant states of <span class="math inline">\(\theta\)</span>.</p>
</div>
<div id="why-so-many-options" class="section level2">
<h2>Why so many options?</h2>
<p>Each <span class="math inline">\(\theta\)</span> type serves a different research purpose. The <span class="math inline">\(\theta_d\)</span> measure tends to identify pairs with a large difference in their log-ratio means. The <span class="math inline">\(\theta_e\)</span> measure tends to identify pairs with a large difference in their VLR. Like <span class="math inline">\(\theta_d\)</span>, <span class="math inline">\(\theta_f\)</span> tends to identify pairs with a large difference in their log-ratio means, but these pairs also tend to have small differences in their VLR.</p>
<p>Each <span class="math inline">\(\theta\)</span> can exist in any of the four variant states. In some contexts, the use of a variant can augment an analysis. For example, the <span class="math inline">\(\alpha\)</span> transformation provides a way of computing VLR in the presence of zeros (although missing data may still require prior imputing). Meanwhile, precision weights provide a way of using information about the distribution of the feature counts to adjust VLR estimates, shrinking the VLR for pairs with lowly abundant features.</p>
</div>
<div id="calculating-the-f-statistic" class="section level2">
<h2>Calculating the F-statistic</h2>
<p>In the companion vignette, we show how it is possible to use the <code>updateCutoffs</code> function to calculate a false discovery rate (FDR) using permutations of the group assignments to generate an empiric null distribution of <span class="math inline">\(\theta\)</span> values. This method works for all <span class="math inline">\(\theta\)</span> types and variant states.</p>
<p>However, analysts may find it desirable or necessary to calculate a <span class="math inline">\(p\)</span>-value exactly. For this, we provide a relationship between <span class="math inline">\(\theta_d\)</span> and the <span class="math inline">\(F\)</span>-statistic:</p>
<p><span class="math display">\[F = (n-2)\frac{1-\theta_d}{\theta_d}\]</span></p>
<p>This relationship holds true regardless of whether precision weights or a power transformation is used. The <code>updateF</code> function calculates the <span class="math inline">\(F\)</span>-statistic from <span class="math inline">\(\theta_d\)</span>, appending an “Fstat” column to the <code>@theta</code> slot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pd.nn &lt;-<span class="st"> </span><span class="kw">updateF</span>(pd.nn, <span class="dt">moderated =</span> <span class="ot">FALSE</span>)
pd.nn@theta$Fstat</code></pre></div>
</div>
<div id="moderating-the-f-statistic" class="section level2">
<h2>Moderating the F-statistic</h2>
<p>Borrowing again from the <code>limma</code> package, we offer a way to calculate a moderated <span class="math inline">\(F\)</span>-statistic for differential proportionality analysis. The principle behind moderation states that it is possible to “borrow information between genes” (i.e., via a Bayesian hierarchical model) to improve the power of statistical hypothesis testing in the setting of small sample sizes (Smyth 2004). This technique was first for developed for measuring the differential expression (DE) of (normally distributed) microarray data, but was subsequently extended to RNA-Seq count data through the use of precision weights (Law 2014). Precision weights, like those described above, can model mean-variance trends in count data to facilitate the analysis of counts as if they were normally distributed (Law 2014).</p>
<p>To calculate a moderated <span class="math inline">\(F\)</span>-statistic from <span class="math inline">\(\theta_d\)</span>, we fit the data to an empirical Bayes model (via <code>limma::eBayes</code>) with underlying mean-variance modeling (via <code>limma::voom</code>). Conventionally, for per-gene (i.e., DE) analysis (and also for VLR weighting), moderation and modeling is done for individual genes. However, for per-ratio (i.e., <span class="math inline">\(\theta_d\)</span>) analysis, moderation and modeling is done for gene ratios. To apply the per-gene moderation to ratios, we must select a suitable reference, <span class="math inline">\(z\)</span>, which is used for a kind of normalization of the data. The hierarchical model is then calculated after this normalization is performed. As a consequence, the moderation of the <span class="math inline">\(F\)</span>-statistic depends on the chosen reference (although the unmoderated <span class="math inline">\(F\)</span>-statistic does not).</p>
<p>By default, the reference for each i-th composition (i.e., sample vector) is the geometric mean of the composition itself. Using this reference, our normalization becomes the corresponding log-ratio transformation (i.e., the clr transformation in the default case). However, <code>limma::voom</code> adds pseudo-counts to the normalized counts, and in order for these to have a similar influence as if applied to the original counts, we must up-scale our log-ratio transformed data again. We achieve this by multiplying (the exponential of) the log-ratio transformed data by the mean over the reference (i.e., the constant <span class="math inline">\(\sum_i^N \mathbf{z}_i/N\)</span>). It is this product that is used to obtain the hierarchical model parameters. In contrast, the weights themselves are calculated directly from the original counts.</p>
<p>The <code>updateF</code> function calculates the moderated <span class="math inline">\(F\)</span>-statistic if the argument <code>moderated</code> equals <code>TRUE</code>. Meanwhile, the <code>ivar</code> argument defines the arbitrary feature set to use as the reference. By default, the <code>updateF</code> function uses the geometric mean of all features as the reference (analogous to the clr transformation), although the user may specify any reference as described in <code>?updateF</code>. This function appends an “Fstat” and “theta_mod” column to the <code>@theta</code> slot. Note that while per-ratio modeling is used to moderate the <span class="math inline">\(F\)</span>-statistic, per-gene modeling is still used to calculate a weighted VLR. Although <code>limma::voom</code> is used in both scenarios, it remains possible to moderate the <span class="math inline">\(F\)</span>-statistic without weighting VLR (and <em>vice versa</em>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pd.nn &lt;-<span class="st"> </span><span class="kw">updateF</span>(pd.nn, <span class="dt">moderated =</span> <span class="ot">TRUE</span>, <span class="dt">ivar =</span> <span class="st">&quot;clr&quot;</span>)
pd.wn &lt;-<span class="st"> </span><span class="kw">updateF</span>(pd.wn, <span class="dt">moderated =</span> <span class="ot">TRUE</span>, <span class="dt">ivar =</span> <span class="st">&quot;clr&quot;</span>)
pd.na &lt;-<span class="st"> </span><span class="kw">updateF</span>(pd.na, <span class="dt">moderated =</span> <span class="ot">TRUE</span>, <span class="dt">ivar =</span> <span class="st">&quot;clr&quot;</span>)
pd.wa &lt;-<span class="st"> </span><span class="kw">updateF</span>(pd.wa, <span class="dt">moderated =</span> <span class="ot">TRUE</span>, <span class="dt">ivar =</span> <span class="st">&quot;clr&quot;</span>)</code></pre></div>
<p>We refer the reader to Erb et al. 2017 for an elaboration of <span class="math inline">\(F\)</span>-statistic moderation.</p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ol style="list-style-type: decimal">
<li><p>Erb, Ionas, Thomas Quinn, David Lovell, and Cedric Notredame. “Differential Proportionality - A Normalization-Free Approach To Differential Gene Expression.” bioRxiv, May 5, 2017, 134536. <a href="http://dx.doi.org/10.1101/134536" class="uri">http://dx.doi.org/10.1101/134536</a>.</p></li>
<li><p>Law, Charity W., Yunshun Chen, Wei Shi, and Gordon K. Smyth. “Voom: Precision Weights Unlock Linear Model Analysis Tools for RNA-Seq Read Counts.” Genome Biology 15 (January 3, 2014): R29. <a href="https://doi.org/10.1186/gb-2014-15-2-r29" class="uri">https://doi.org/10.1186/gb-2014-15-2-r29</a>.</p></li>
<li><p>Smyth, Gordon K. “Linear Models and Empirical Bayes Methods for Assessing Differential Expression in Microarray Experiments.” Statistical Applications in Genetics and Molecular Biology 3 (2004): Article3. <a href="https://doi.org/10.2202/1544-6115.1027" class="uri">https://doi.org/10.2202/1544-6115.1027</a>.</p></li>
</ol>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
