<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Thomas Quinn" />

<meta name="date" content="2017-08-01" />

<title>Frequently Asked Questions</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Frequently Asked Questions</h1>
<h4 class="author"><em>Thomas Quinn</em></h4>
<h4 class="date"><em>2017-08-01</em></h4>



<div id="why-does-proportionality-change-when-i-remove-features" class="section level2">
<h2>Why does proportionality change when I remove features?</h2>
<p>We use the term sub-compositional coherence to describe a method in which the results do not change when features (e.g., genes) are added or removed. The numerator portion of <span class="math inline">\(\rho\)</span> (i.e., the VLR) is sub-compositionally coherent. However, <span class="math inline">\(\rho\)</span> itself is not sub-compositionally coherent because the denominator portion of <span class="math inline">\(\rho\)</span> (and <span class="math inline">\(\phi\)</span> and <span class="math inline">\(\phi_s\)</span>) scales the VLR by the variance of the centered log-ratio (clr) transformed features. The clr-transformation is not sub-compositionally coherent because the removal of a feature will change geometric mean reference for each composition (i.e., each sample) and therefore its transformation; consequently, the variance of each transformed feature vector changes too. An exception to this is when using the additive log-ratio (alr) transformation instead of clr.</p>
</div>
<div id="isnt-the-centered-log-ratio-bad-then" class="section level2">
<h2>Isn’t the centered log-ratio bad then?</h2>
<p>The centered log-ratio is not perfect, but it is maybe <em>good enough</em>. Still, this is the major limitation of proportionality analysis: it inherits all the limitations of the choice in log-ratio transformation. However, as long as the majority of genes remain unchanged across the conditions, the clr is appropriate. Excerpted from Erb et al. 2016:</p>
<blockquote>
<p>This is known as the centered log-ratio transformation with g(x) the geometric mean over the genes for the given condition. The problem with this transformation is that it is sub-compositionally incoherent Aitchison (2003), so results will change to some extent when using subsets of genes for the analysis. In some cases, g(x) can approximate an unchanged reference. This applies whenever the majority of genes remains unchanged across conditions, so the unchanged genes will dominate the behaviour of the reference. Note that this is also the condition needed for a normalization by effective library size Robinson and Oshlack (2012).</p>
</blockquote>
</div>
<div id="what-if-the-centered-log-ratio-isnt-for-me" class="section level2">
<h2>What if the centered log-ratio isn’t for me?</h2>
<p>One option is to use a house-keeping gene or RNA spike-in that has an <em>a priori</em> known fixed abundance. If this is available, you can use the additive log-ratio (alr) transformation to effectively “back-calculate” absolute abundances. In practice, this is sometimes infeasible. Another option is to use a novel transformation introduced in Fernandes et al. 2013 called the inter-quartile log-ratio (iqlr) transformation. Excerpted from the <code>ALDEx2</code> vignette:</p>
<blockquote>
<p>The [approach] is to include as the denominator for the geometric mean those features that are relatively invariant across all samples. This is termed the ‘iqlr’ method, and takes as the denominator the geometric mean of those features with variance calculated from the clr that are between the first and third quartile. This approach can be used until the asymmetry becomes so severe that more than 25% of the features are asymmetric between the groups. The iqlr approach has the advantage that it gives essentially the same answer as using the entire set of features in symmetric datasets.</p>
</blockquote>
<p>Greg Gloor and I have coordinated to make the iqlr transformation and the Monte Carlo (MC) instances from <code>ALDEx2</code> available for proportionality analysis in <code>propr</code>. You can use the <code>aldex2propr</code> function to build a <code>propr</code> object from an <code>aldex.clr</code> object. In practice, the code might look something like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(caneToad.counts)
<span class="kw">data</span>(caneToad.groups)
counts &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(caneToad.counts))
x &lt;-<span class="st"> </span>ALDEx2::<span class="kw">aldex.clr</span>(counts, caneToad.groups, <span class="dt">denom =</span> <span class="st">&quot;iqlr&quot;</span>)
rho &lt;-<span class="st"> </span>propr::<span class="kw">aldex2propr</span>(x, <span class="dt">how =</span> <span class="st">&quot;rho&quot;</span>)</code></pre></div>
<p>Note that this will average <span class="math inline">\(\rho\)</span> (or <span class="math inline">\(\phi\)</span> or <span class="math inline">\(\phi_s\)</span>) across many MC instances which may or may not be desired. In theory, using MC instances should lessen the impact of zero or low counts on your final result. In practice, proportionality analysis will take a lot longer to run. We have not yet benchmarked whether <code>aldex2propr</code> is better or worse than <code>perb</code> alone, but find the iqlr transformation a very attractive option. Note, however, that you can also use the iqlr transformation without any MC instances:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rho &lt;-<span class="st"> </span><span class="kw">perb</span>(caneToad.counts, <span class="dt">ivar =</span> <span class="st">&quot;iqlr&quot;</span>)</code></pre></div>
<p>When using <code>aldex2propr</code> (or the iqlr transformation), please make sure to cite the appropriate <code>ALDEx2</code> paper(s) to thank Greg for his generous contribution!</p>
</div>
<div id="why-cant-we-just-use-the-vlr-by-itself" class="section level2">
<h2>Why can’t we just use the VLR by itself?</h2>
<p>Quoting Lovell et al. 2015 (quoting Friedman et al. 2012):</p>
<blockquote>
<p>Aitchison proposed logratio variance, var(log(x/y)), as a measure of association for variables that carry only relative information. When x and y are exactly proportional var(log(x/y)) = 0, but when x and y are not exactly proportional, “it is hard to interpret as it lacks a scale. That is, it is unclear what constitutes a large or small value… (does a value of 0.1 indicate strong dependence, weak dependence, or no dependence?)”</p>
</blockquote>
</div>
<div id="why-does-my-smear-plot-show-perfectly-proportional-pairs" class="section level2">
<h2>Why does my <code>smear</code> plot show perfectly proportional pairs?</h2>
<p>We call this phenomena <strong>auto-proportionality</strong>. Depending on how genes (or transcripts) get mapped, annotations with similar open reading frames (ORFs) can wrongly appear as if they have proportional abundance only because some sequences map to multiple annotations equally well. This is especially a problem when analyzing transcriptomic data: transcript isoforms tend to appear proportional with one another by virtue of their shared ORF. The discovery of proportional isoforms could reflect a real biological event (i.e., transcript isoforms under shared regulatory control) or an artifact of mapping. We recommend removing such pairs to avoid swamping interesting biological signals with artifactual noise. You can do this by removing the set of extremely proportional pairs (e.g., <span class="math inline">\(\rho &gt; .995\)</span>) from the set of highly proportional pairs (e.g., <span class="math inline">\(\rho &gt; .95\)</span>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(caneToad.counts)
rho &lt;-<span class="st"> </span><span class="kw">perb</span>(caneToad.counts)
autoprop &lt;-<span class="st"> </span>rho[<span class="st">&quot;&gt;&quot;</span>, .<span class="dv">995</span>]@pairs
highprop &lt;-<span class="st"> </span>rho[<span class="st">&quot;&gt;&quot;</span>, .<span class="dv">95</span>]@pairs
rho@pairs &lt;-<span class="st"> </span><span class="kw">setdiff</span>(highprop, autoprop)</code></pre></div>
</div>
<div id="how-can-i-analyze-multi-omics-data" class="section level2">
<h2>How can I analyze multi-omics data?</h2>
<p>This is an area of active research. At a minimum, we recommend making sure that (a) all combined multi-omics data derive from the same samples and (b) log-ratio transformations are applied to each data source individually. The functions <code>propr:::lr2rho</code>, <code>propr:::lr2phi</code>, and <code>propr:::lr2phs</code>, used in conjunction with <code>new(&quot;propr&quot;)</code>, offer a way forward. The <code>propr</code> initialization method will perform the log-ratio transformation without building a proportionality matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(caneToad.counts)
prop &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;propr&quot;</span>, <span class="dt">counts =</span> caneToad.counts, <span class="dt">ivar =</span> <span class="st">&quot;iqlr&quot;</span>)
prop@logratio</code></pre></div>
<p>You could then <code>cbind</code> any number of separate <code>@logratio</code> slots into a single <code>data.frame</code> to use as input for an “lr2” function.</p>
</div>
<div id="how-can-i-analyze-multiple-groups" class="section level2">
<h2>How can I analyze multiple groups?</h2>
<p>This is another area of active research. At a minimum, we recommend making sure that data filtering (e.g., removal of low and zero counts) and log-ratio transformation is applied identically to all data prior to analysis. From there, the usual pipeline would likely involve separating the data into several pairwise (i.e., two-way) comparisons. Users should take care when crafting their hypotheses and consider whether one-vs-all or one-vs-one comparisons better suit their goals.</p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ol style="list-style-type: decimal">
<li><p>Erb, Ionas, and Cedric Notredame. “How Should We Measure Proportionality on Relative Gene Expression Data?” Theory in Biosciences = Theorie in Den Biowissenschaften 135, no. 1-2 (June 2016): 21-36. <a href="http://dx.doi.org/10.1007/s12064-015-0220-8" class="uri">http://dx.doi.org/10.1007/s12064-015-0220-8</a>.</p></li>
<li><p>Fernandes, Andrew D., Jean M. Macklaim, Thomas G. Linn, Gregor Reid, and Gregory B. Gloor. “ANOVA-like Differential Expression (ALDEx) Analysis for Mixed Population RNA-Seq.” PloS One 8, no. 7 (2013): e67019. <a href="http://dx.doi.org/10.1371/journal.pone.0067019" class="uri">http://dx.doi.org/10.1371/journal.pone.0067019</a>.</p></li>
<li><p>Friedman, Jonathan, and Eric J. Alm. “Inferring Correlation Networks from Genomic Survey Data.” PLoS Computational Biology 8, no. 9 (2012): e1002687. <a href="http://dx.doi.org/10.1371/journal.pcbi.1002687" class="uri">http://dx.doi.org/10.1371/journal.pcbi.1002687</a>.</p></li>
<li><p>Lovell, David, Vera Pawlowsky-Glahn, Juan José Egozcue, Samuel Marguerat, and Jürg Bähler. “Proportionality: A Valid Alternative to Correlation for Relative Data.” PLoS Computational Biology 11, no. 3 (March 2015): e1004075. <a href="http://dx.doi.org/10.1371/journal.pcbi.1004075" class="uri">http://dx.doi.org/10.1371/journal.pcbi.1004075</a>.</p></li>
</ol>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
